probe-rs/src/architecture/arm/memory/adi_v5_memory_interface.rs

```diff
@@ -8,23 +8,95 @@ use crate::architecture::arm::{
 };
 use crate::{CommunicationInterface, Error};
-use scroll::{Pread, Pwrite, LE};
 use std::convert::TryInto;
 use std::ops::Range;
 
 pub trait ArmProbe: SwdSequence {
     fn read_8(&mut self, ap: MemoryAp, address: u64, data: &mut [u8]) -> Result<(), Error>;
+
     fn read_32(&mut self, ap: MemoryAp, address: u64, data: &mut [u32]) -> Result<(), Error>;
+
     fn read_64(&mut self, ap: MemoryAp, address: u64, data: &mut [u64]) -> Result<(), Error>;
 
+    /// Read a block of 8bit words at `address`. May use 32 bit memory access,
+    /// so should only be used if reading memory locations that don't have side
+    /// effects. Generally faster than [`MemoryInterface::read_8`].
+    fn read(&mut self, ap: MemoryAp, address: u64, data: &mut [u8]) -> Result<(), Error> {
+        let len = data.len();
+        if address % 4 == 0 && len % 4 == 0 {
+            let mut buffer = vec![0u32; len / 4];
+            self.read_32(ap, address, &mut buffer)?;
+            for (bytes, value) in data.chunks_exact_mut(4).zip(buffer.iter()) {
+                bytes.copy_from_slice(&u32::to_le_bytes(*value));
+            }
+        } else {
+            let start_extra_count = (address % 4) as usize;
+            let mut buffer = vec![0u32; (start_extra_count + len + 3) / 4];
+            let read_address = address - start_extra_count as u64;
+            self.read_32(ap, read_address, &mut buffer)?;
+            for (bytes, value) in data
+                .chunks_exact_mut(4)
+                .zip(buffer[start_extra_count..start_extra_count + len].iter())
+            {
+                bytes.copy_from_slice(&u32::to_le_bytes(*value));
+            }
+        }
+        Ok(())
+    }
+
     fn write_8(&mut self, ap: MemoryAp, address: u64, data: &[u8]) -> Result<(), Error>;
+
     fn write_32(&mut self, ap: MemoryAp, address: u64, data: &[u32]) -> Result<(), Error>;
+
     fn write_64(&mut self, ap: MemoryAp, address: u64, data: &[u64]) -> Result<(), Error>;
 
+    /// Write a block of 8bit words to `address`. May use 32 bit memory access,
+    /// so it should only be used if writing memory locations that don't have side
+    /// effects. Generally faster than [`MemoryInterface::write_8`].
+    fn write(&mut self, ap: MemoryAp, address: u64, data: &[u8]) -> Result<(), Error> {
+        let len = data.len();
+        let start_extra_count = 4 - (address % 4) as usize;
+        let end_extra_count = ((len - start_extra_count) % 4) as usize;
+        let inbetween_count = len - start_extra_count - end_extra_count;
+        assert!(start_extra_count < 4);
+        assert!(end_extra_count < 4);
+        assert!(inbetween_count % 4 == 0);
+
+        // If we do not have 32 bit aligned access we first check that we can do 8 bit aligned access on this platform.
+        // If we cannot we throw an error.
+        // If we can we read the first n < 4 bytes up until the word aligned address that comes next.
+        if address % 4 != 0 && len % 4 != 0 {
```

Comment:
Shouldn't this be an or? If the address is aligned, but the length is not a multiple of 4, using only 32 bit accesses is also not possible